main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala

```diff
@@ -0,0 +1,3144 @@
+    }
+
+    private def visitLiteralArray(tree: Tree)(implicit s: State): Validation[Expr, CompilationMessage] = {
+      assert(tree.kind == TreeKind.Expr.LiteralArray)
+      val exprs = pickAll(TreeKind.Expr.Expr, tree)
+      val scopeName = tryPick(TreeKind.Expr.ScopeName, tree)
+      flatMapN(
+        traverse(exprs)(visitExpression),
+        traverseOpt(scopeName)(visitScopeName)
+      ) {
+        case (exprs, Some(scope)) => Validation.success(Expr.ArrayLit(exprs, scope, tree.loc))
+        case (exprs, None) =>
+          val err = ParseError("Missing scope in array literal.", SyntacticContext.Expr.OtherExpr, tree.loc)
+          Validation.toSoftFailure(Expr.ArrayLit(exprs, Expr.Error(err), tree.loc), err)
+      }
+    }
+
+    private def visitLiteralMap(tree: Tree)(implicit s: State): Validation[Expr, CompilationMessage] = {
+      assert(tree.kind == TreeKind.Expr.LiteralMap)
+      val pairs = pickAll(TreeKind.Expr.LiteralMapKeyValueFragment, tree)
+      mapN(traverse(pairs)(visitKeyValuePair))(Expr.MapLit(_, tree.loc))
+    }
+
+    private def visitKeyValuePair(tree: Tree)(implicit s: State): Validation[(Expr, Expr), CompilationMessage] = {
+      assert(tree.kind == TreeKind.Expr.LiteralMapKeyValueFragment)
+      val exprs = pickAll(TreeKind.Expr.Expr, tree)
+      flatMapN(traverse(exprs)(visitExpression)) {
+        // case: k => v
+        case k :: v :: Nil => Validation.success((k, v))
+        // case: k =>
+        case k :: Nil =>
+          val err = ParseError("Missing value in key-value pair", SyntacticContext.Expr.OtherExpr, tree.loc)
+          Validation.toSoftFailure((k, Expr.Error(err)), err)
+        case xs => throw InternalCompilerException(s"Malformed KeyValue pair, found ${xs.length} expressions", tree.loc)
+      }
+    }
+
+    private def visitScope(tree: Tree)(implicit s: State): Validation[Expr, CompilationMessage] = {
+      assert(tree.kind == TreeKind.Expr.Scope)
+      val block = flatMapN(pick(TreeKind.Expr.Block, tree))(visitBlock)
+      mapN(pickNameIdent(tree), block)((ident, block) => Expr.Scope(ident, block, tree.loc))
+    }
+
+    private def tryPickNumberLiteralToken(tree: Tree)(implicit s: State): Option[Token] = {
+      val NumberLiteralKinds = List(TokenKind.LiteralInt8, TokenKind.LiteralInt16, TokenKind.LiteralInt32, TokenKind.LiteralInt64, TokenKind.LiteralBigInt, TokenKind.LiteralFloat32, TokenKind.LiteralFloat64, TokenKind.LiteralBigDecimal)
+      val maybeTree = tryPick(TreeKind.Expr.Literal, tree)
+      maybeTree.flatMap(_.children(0) match {
+        case Child.TokenChild(t) if NumberLiteralKinds.contains(t.kind) => Some(t)
+        case _ => None
+      })
+    }
+
+    private def visitUnary(tree: Tree)(implicit s: State): Validation[Expr, CompilationMessage] = {
+      assert(tree.kind == TreeKind.Expr.Unary)
+      flatMapN(pick(TreeKind.Operator, tree), pick(TreeKind.Expr.Expr, tree))(
+        (opTree, exprTree) => {
+          opTree.children(0) match {
```

Comment:
How do we know this is within bounds?

Reply:
The parser guarantees that it will never produce a tree of kind `TreeKind.Op` without a child. I changed it to `headOption` though, since an implicit guarantee like that is confusing.